lib/app_profiler/server.rb

```diff
@@ -0,0 +1,292 @@
+          response.status = HTTP_NOT_ALLOWED
+          response.write("Only GET requests are supported")
+          return response
+        end
+        case request.path
+        when "/profile"
+          begin
+            stackprof_args, duration = validate_profile_params(request.params)
+          rescue => e
+            response.status = HTTP_BAD_REQUEST
+            response.write("Invalid argument #{e.message}")
+            return response
+          end
+
+          if start_running
+            start_time = Process.clock_gettime(Process::CLOCK_REALTIME, :nanosecond)
+            AppProfiler.start(**stackprof_args)
+            sleep(duration)
+            profile = AppProfiler.stop
+            stop_running
+            response.status = HTTP_OK
+            response.set_header("Content-Type", "application/json")
+            if AppProfiler::Server.cors
+              response.set_header("Access-Control-Allow-Origin", AppProfiler::Server.cors_host)
+            end
+            profile_hash = profile.to_h
+            profile_hash["start_time_nsecs"] = start_time # NOTE: this is not part of the stackprof profile spec
+            response.write(JSON.dump(profile_hash))
+          else
+            response.status = HTTP_CONFLICT
+            response.write("A profile is already running")
+          end
+        else
+          response.status = HTTP_NOT_FOUND
+          response.write("Unsupported endpoint #{request.path}")
+        end
+        response
+      end
+
+      def validate_profile_params(params)
+        params = params.symbolize_keys
+        stackprof_args = {}
+        duration = Float(params.key?(:duration) ? params[:duration] : AppProfiler::Server.duration)
+        if params.key?(:mode)
+          if ["cpu", "wall", "object"].include?(params[:mode])
+            stackprof_args[:mode] = params[:mode].to_sym
+          else
+            raise InvalidProfileArgsError, "invalid mode #{params[:mode]}"
+          end
+        end
+        if params.key?(:interval)
+          stackprof_args[:interval] = params[:interval].to_i
+          raise InvalidProfileArgsError, "invalid interval #{params[:interval]}" if stackprof_args[:interval] <= 0
+        end
+        [stackprof_args, duration]
+      end
+
+      # Prevent multiple concurrent profiles by synchronizing between threads
+      def start_running
+        @semaphore.synchronize do
+          return false if @profile_running
+
+          @profile_running = true
+        end
+      end
+
+      def stop_running
+        @semaphore.synchronize { @profile_running = false }
+      end
+    end
+
+    # This is a minimal, non-compliant "HTTP" server.
+    # It will create an extremely minimal rack environment hash and hand it off
+    # to our application to process
+    class ProfileServer
+      PROFILER_TEMPFILE_PATH = "/tmp/app_profiler" # for tempfile that indicates port in filename or unix sockets
+
+      class ProfileServerTransport
```

Comment:
```suggestion
      class Transport
```
This is already within the `ProfileServer` module, so `ProfileServer::Transport` makes more sense to me. It also allows for more readable subclasses: `class UNIX < Transport`.